### PR TITLE
Replace editoast JsonValue by appropriate structs

### LINKED ISSUE #5457: editoast: replace serde_json::Value occurrences by actual structs where appropriate
### Description and goal

* Remove the `serde_json::Value` usage where serializable structs should be used (some are defined, some you will have to define).
  * To define the missing structs, you can either look at the openAPI or at the python `osrd-schemas` library.
*  Here are the areas where `serde_json::Value` is still used where it shouldn't (I might have missed some 🙈)
    * in the models RollingStock and LightRollingStock, for power restrictions
    * in the schemas RollingStock and LightRollingStock, for power restrictions
* ~some `serde_json::Value` remain in /infra related views, but that's expected since they're yet to be annotated with utoipa~

### Acceptance criteria

- [ ] No `serde_json::Value` is left

### PR DESCRIPTION
**Pull request by hamz2a**
closes #5457

### CODE REVIEW

File: editoast/src/schema/errors.rs [hunk 1]
```diff
@@ -4,14 +4,13 @@ use serde::{Deserialize, Serialize};
...
 #[derive(Serialize, Deserialize, PartialEq, Debug, Clone)]
-#[serde(deny_unknown_fields)]
 pub struct InfraError {
     obj_id: String,
     obj_type: ObjectType,
     field: Option<String>,
     is_warning: bool,
     #[serde(flatten)]
-    sub_type: InfraErrorType,
+    error_type: InfraErrorType,
```
Comment (flomonster) [inline editoast/src/schema/errors.rs]: We can not rename this since this struct is used to deserialize data. Do we have reason to want to rename this field? Also, this seems to be outside the scope of the RP.

File: editoast/src/schema/errors.rs [hunk 2]
```diff
@@ -4,14 +4,13 @@ use serde::{Deserialize, Serialize};
 use strum_macros::EnumVariantNames;
 
 #[derive(Serialize, Deserialize, PartialEq, Debug, Clone)]
-#[serde(deny_unknown_fields)]
```
Comment (flomonster) [inline editoast/src/schema/errors.rs]: We want to keep it that way as much as possible. It allows the detection of weird payloads.

File: editoast/src/views/infra/errors.rs [hunk 3]
```diff
@@ -65,10 +64,9 @@ enum ListErrorsErrors {
 }
 
 #[derive(QueryableByName, Debug, Clone, Serialize, Deserialize, PartialEq)]
-#[serde(deny_unknown_fields)]
 pub struct InfraError {
-    #[diesel(sql_type = Json)]
-    pub information: JsonValue,
+    #[diesel(sql_type = Jsonb)]
+    pub information: diesel_json::Json<crate::schema::InfraError>,
```
Comment (flomonster) [inline editoast/src/views/infra/errors.rs]:
```suggestion
    pub information: DieselJson<InfraError>,
```

File: editoast/src/views/infra/objects.rs [hunk 4]
```diff
@@ -155,6 +157,68 @@ mod tests {
...
+                        ],
+                        [
+                            -0.365,
+                            49.5
+                        ]
+                    ],
+                    "type": "LineString"
+                },
+                "slopes": []
+            }),
```
Comment (flomonster) [inline editoast/src/views/infra/objects.rs]: This is highly interrelated with `small_infra.json`. We could untie this by:
- Retrieving the payload of `TA0`
- Using an empty infra and creating our own track section and checking that the endpoint returns it

File: editoast/src/views/layers/mod.rs [hunk 5]
```diff
@@ -75,15 +76,18 @@ struct LayerViewParams {
...
+#[derive(Debug, Serialize, Deserialize, PartialEq, ToSchema)]
 struct ViewMetadata {
     #[serde(rename = "type")]
     data_type: String,
     #[schema(example = "track_sections")]
     name: String,
     #[serde(rename = "promoteId")]
     #[schema(example = json!({track_sections: "id"}), value_type = Object)]
-    promote_id: JsonValue,
+    promote_id: PromoteIdWrapper,
```
Comment (flomonster) [inline editoast/src/views/layers/mod.rs]: You can add `PromoteIdWrapper` to the schema to avoid having `value_type = Object`

**Review by flomonster — CHANGES_REQUESTED** [review]

File: editoast/src/views/infra/routes.rs [hunk 6]
```diff
@@ -119,3 +128,78 @@ async fn get_routes_track_ranges<'a>(
...
+            .uri(
+                format!("/infra/{small_infra_id}/routes/{waypoint_type}/NOT_EXISTING_WAYPOINT_ID")
+                    .as_str(),
+            )
+            .to_request();
+        let response = call_service(&app, request).await;
+        assert_eq!(response.status(), StatusCode::OK);
+        let waypoint: WaypointResponse = read_body_json(response).await;
+        assert_eq!(waypoint, WaypointResponse::new(vec![], vec![]));
+    }
```
Comment (Tguisnet) [inline editoast/src/views/infra/routes.rs]: You could use an empty infra here.

File: editoast/src/views/infra/routes.rs [hunk 7]
```diff
@@ -119,3 +128,78 @@ async fn get_routes_track_ranges<'a>(
...
+    use crate::{
+        fixtures::tests::{db_pool, small_infra},
+        views::{
+            infra::routes::{WaypointResponse, WaypointType},
+            tests::create_test_service,
+        },
+    };
+
+    #[rstest]
+    async fn get_waypoint_should_return_da2_detector() {
```
Comment (Tguisnet) [inline editoast/src/views/infra/routes.rs]: Not sure about the naming, this should not return a detector, but a list of routes from a waypoint.
Also I'm not a big fan of naming test function with an id inside. (here da2)

File: editoast/src/views/infra/routes.rs [hunk 8]
```diff
@@ -119,3 +128,78 @@ async fn get_routes_track_ranges<'a>(
...
+            waypoint,
+            WaypointResponse::new(
+                vec!["rt.DA2->DA5".to_string()],
+                vec!["rt.buffer_stop.0->DA2".to_string()]
+            )
+        );
+    }
+
+    #[rstest]
+    async fn get_waypoint_should_return_buffer_stop_number_4() {
```
Comment (Tguisnet) [inline editoast/src/views/infra/routes.rs]: Not sure about the naming, this should not return a detector, but a list of routes from a waypoint.
Also I'm not a big fan of naming test function with an id inside. (here number 4)

File: editoast/src/schema/errors.rs (hunk 1 above)
Comment (Tguisnet) [inline editoast/src/schema/errors.rs]: Is there a reason for this renaming ?

File: editoast/src/views/infra/objects.rs (hunk 4 above)
Comment (Tguisnet) [inline editoast/src/views/infra/objects.rs]: You could avoid this with the creation of your own track on an empty infra (avoiding the creation of a small infra too)

File: editoast/src/views/infra/routes.rs [hunk 9]
```diff
@@ -36,18 +35,30 @@ enum WaypointType {
...
+
+impl WaypointResponse {
+    pub fn new(starting: Vec<String>, ending: Vec<String>) -> Self {
+        Self { starting, ending }
+    }
+}
+
 /// Return the railjson list of a specific OSRD object
-#[get("/{waypoint_type}/{waypoint}")]
+#[get("/{waypoint_type}/{waypoint_id}")]
```
Comment (Tguisnet) [inline editoast/src/views/infra/routes.rs]: This is not a change you did, but this comment is innacurate and could be confusing.
We should change it.

File: editoast/src/views/infra/routes.rs [hunk 10]
```diff
@@ -36,18 +35,30 @@ enum WaypointType {
...
+struct WaypointResponse {
+    starting: Vec<String>,
+    ending: Vec<String>,
+}
+
+impl WaypointResponse {
+    pub fn new(starting: Vec<String>, ending: Vec<String>) -> Self {
+        Self { starting, ending }
+    }
+}
```
Comment (Tguisnet) [inline editoast/src/views/infra/routes.rs]: Not sure about the naming here.

**Review by Tguisnet — CHANGES_REQUESTED** [review]

File: editoast/src/schema/rolling_stock/mod.rs [hunk 11]
```diff
@@ -52,8 +52,8 @@ pub struct RollingStockCommon {
     #[schema(value_type = LoadingGaugeType)]
     pub loading_gauge: String,
     /// Mapping of power restriction code to power class
-    #[schema(value_type = HashMap<String, String>)]
-    pub power_restrictions: Option<JsonValue>,
+    #[schema(value_type = PowerRestriction)]
```
Comment (leovalais) [inline editoast/src/schema/rolling_stock/mod.rs]: That's not how utoipa works. If you replaced the type by `Lolz` editoast would compile just as well, yielding an invalid ref in the openapi. PowerRestriction happens to exist in our openapi but has a totally different meaning. utoipa has a few unexpected design takes (which are probably justified), and you are hitting one of them. The documentation of the crate and the `utoipa::path`, `utoipa::OpenApi` macros are a good way to have an overview.

File: editoast/openapi.yaml [hunk 12]
```diff
@@ -859,9 +859,7 @@ components:
         name:
           type: string
         power_restrictions:
-          additionalProperties:
-            type: string
-          type: object
+          $ref: '#/components/schemas/PowerRestriction'
```
Comment (leovalais) [inline editoast/openapi.yaml]: Wrong schema

File: editoast/src/models/rolling_stock/mod.rs [hunk 13]
```diff
@@ -109,9 +108,9 @@ pub struct RollingStockModel {
     #[diesel(deserialize_as = String)]
     #[schema(value_type = LoadingGaugeType)]
     pub loading_gauge: Option<String>,
-    #[diesel(deserialize_as = Option<JsonValue>)]
-    #[schema(value_type = Option<HashMap<String, String>>)]
-    pub power_restrictions: Option<Option<JsonValue>>,
+    #[diesel(deserialize_as = Option<DieselJson<PowerRestriction>>)]
+    #[schema(value_type = Option<PowerRestriction>)]
```
Comment (leovalais) [inline editoast/src/models/rolling_stock/mod.rs]: It's not a truly optional field, `Option<Option<T>>` are optional fields in the `Model` context.

File: editoast/src/schema/errors.rs (hunk 1 above)
Comment (hamz2a) [reply editoast/src/schema/errors.rs]: This field is stored in the database under the name `error_type`.

File: editoast/src/views/layers/mod.rs [hunk 14]
```diff
@@ -75,15 +76,18 @@ struct LayerViewParams {
     view_slug: String,
 }
 
-#[derive(Serialize, ToSchema)]
+#[derive(Debug, Serialize, Deserialize, PartialEq)]
+struct PromoteIdWrapper(HashMap<String, String>);
```
Comment (leovalais) [inline editoast/src/views/layers/mod.rs]: You need to provide more info, check out `utoipa::ToSchema` docs and the macro `schemas!`

**Review by leovalais — CHANGES_REQUESTED** [review]
Thanks for this PR. There are issues with the OpenApi annotations and generation, more info in the comments ⬇️

File: editoast/src/schema/errors.rs (hunk 2 above)
Comment (hamz2a) [reply editoast/src/schema/errors.rs]: `#[serde(deny_unknown_fields)]` if I keep this, deserialize does not work correctly for some tests

File: editoast/src/views/infra/routes.rs (hunk 9 above)
Comment (hamz2a) [reply editoast/src/views/infra/routes.rs]: I have changed the `waypoint` by `waypoint_id`

File: editoast/src/schema/errors.rs (hunk 1 above)
Comment (hamz2a) [reply editoast/src/schema/errors.rs]: This field is stored in the database under the name `error_type`.

File: editoast/src/views/infra/routes.rs (hunk 10 above)
Comment (hamz2a) [reply editoast/src/views/infra/routes.rs]: I changed it by `RoutesResponse`

File: editoast/src/views/infra/routes.rs (hunk 8 above)
Comment (hamz2a) [reply editoast/src/views/infra/routes.rs]: Done.

File: editoast/src/views/infra/routes.rs (hunk 7 above)
Comment (hamz2a) [reply editoast/src/views/infra/routes.rs]: Done.

File: editoast/src/views/infra/routes.rs (hunk 6 above)
Comment (hamz2a) [reply editoast/src/views/infra/routes.rs]: Done.

File: editoast/src/views/infra/routes.rs (hunk 9 above)
Comment (Tguisnet) [reply editoast/src/views/infra/routes.rs]: Yeah, I did not speak about this changes, but about this comment : 
`/// Return the railjson list of a specific OSRD object`

Comment (hamz2a) [reply editoast/src/views/infra/routes.rs]: Ok, do you want me to change it? or just remove it? Any suggestions?

File: editoast/src/views/layers/mod.rs (hunk 14 above)
Comment (hamz2a) [reply editoast/src/views/layers/mod.rs]: I use `HashMap<String, String>` directly in `ViewMetadata`, then we won't need this wrapper anymore.

File: editoast/src/schema/rolling_stock/mod.rs (hunk 11 above)
Comment (hamz2a) [reply editoast/src/schema/rolling_stock/mod.rs]: I use `HashMap<String, String>` directly in `LightRollingStockModel`, `RollingStockModel`, `LightRollingStock` and `RollingStockCommon`, then we won't need this wrapper anymore.

File: editoast/src/views/infra/objects.rs (hunk 4 above)
Comment (hamz2a) [reply editoast/src/views/infra/objects.rs]: Done.

Comment (hamz2a) [reply editoast/src/views/infra/objects.rs]: Done.

File: editoast/src/models/rolling_stock/light_rolling_stock.rs [hunk 15]
```diff
@@ -43,8 +43,8 @@ pub struct LightRollingStockModel {
     rolling_resistance: DieselJson<RollingResistance>,
     #[schema(value_type = LoadingGaugeType)]
     loading_gauge: String,
-    #[schema(value_type = HashMap<String, String>)]
-    power_restrictions: Option<JsonValue>,
+    #[schema(value_type = Option<HashMap<String, String>>)]
```
Comment (leovalais) [inline editoast/src/models/rolling_stock/light_rolling_stock.rs]: This change should be reflected in the openapi, can you check this model is indeed included in the schemas?

File: editoast/src/models/rolling_stock/mod.rs [hunk 16]
```diff
@@ -109,9 +109,9 @@ pub struct RollingStockModel {
     #[diesel(deserialize_as = String)]
     #[schema(value_type = LoadingGaugeType)]
     pub loading_gauge: Option<String>,
-    #[diesel(deserialize_as = Option<JsonValue>)]
+    #[diesel(deserialize_as = Option<DieselJson<HashMap<String, String>>>)]
     #[schema(value_type = Option<HashMap<String, String>>)]
-    pub power_restrictions: Option<Option<JsonValue>>,
+    pub power_restrictions: Option<DieselJson<HashMap<String, String>>>,
```
Comment (leovalais) [inline editoast/src/models/rolling_stock/mod.rs]:
```suggestion
    pub power_restrictions: Option<Option<DieselJson<HashMap<String, String>>>>,
```

File: editoast/src/schema/rolling_stock/light_rolling_stock.rs [hunk 17]
```diff
@@ -61,7 +60,7 @@ pub struct LightRollingStock {
     pub metadata: DieselJson<RollingStockMetadata>,
     #[diesel(sql_type = Nullable<Jsonb>)]
```
Comment (leovalais) [inline editoast/src/schema/rolling_stock/light_rolling_stock.rs:61]: This needs to be updated

File: editoast/src/views/infra/errors.rs [hunk 18]
```diff
@@ -65,10 +66,9 @@ enum ListErrorsErrors {
 }
 
 #[derive(QueryableByName, Debug, Clone, Serialize, Deserialize, PartialEq)]
-#[serde(deny_unknown_fields)]
 pub struct InfraError {
-    #[diesel(sql_type = Json)]
-    pub information: JsonValue,
+    #[diesel(sql_type = Jsonb)]
```
Comment (leovalais) [inline editoast/src/views/infra/errors.rs]: You can be more precise here with DieselJson

File: editoast/src/views/infra/routes.rs [hunk 19]
```diff
@@ -36,18 +35,30 @@ enum WaypointType {
     BufferStop,
 }
 
+#[derive(Debug, Serialize, Deserialize, PartialEq)]
+struct RoutesResponse {
```
Comment (leovalais) [inline editoast/src/views/infra/routes.rs:39]:
```suggestion
struct RoutesResponse {
```

File: editoast/src/views/infra/routes.rs [hunk 20]
```diff
@@ -36,18 +35,30 @@ enum WaypointType {
...
+    starting: Vec<String>,
+    ending: Vec<String>,
+}
+
+impl RoutesResponse {
+    pub fn new(starting: Vec<String>, ending: Vec<String>) -> Self {
+        Self { starting, ending }
+    }
+}
+
```
Comment (leovalais) [inline editoast/src/views/infra/routes.rs]: This `new` is unnecessary since it only does assignment and the struct is defined in the same mod than its usages.

File: editoast/src/views/infra/routes.rs (hunk 9 above)
Comment (leovalais) [reply editoast/src/views/infra/routes.rs]: Better change it.

**Review by leovalais — CHANGES_REQUESTED** [review]

File: editoast/src/schema/errors.rs (hunk 2 above)
Comment (leovalais) [reply editoast/src/schema/errors.rs]: In this case it's the test that should be fixed.

File: editoast/src/schema/errors.rs (hunk 1 above)
Comment (leovalais) [reply editoast/src/schema/errors.rs]: Agreed on not renaming this field. This is tightly tied to auto fixes. If renaming this field is deemed useful, it should be a topic for another PR. And it's not a model, errors in the database are managed by raw SQL, so the field name doesn't necessarily have to match.

File: editoast/src/models/rolling_stock/light_rolling_stock.rs (hunk 15 above)
Comment (hamz2a) [reply editoast/src/models/rolling_stock/light_rolling_stock.rs]:
```
 power_restrictions:
    additionalProperties:
      type: string
    type: object
```

File: editoast/src/schema/rolling_stock/light_rolling_stock.rs (hunk 17 above)
Comment (hamz2a) [reply editoast/src/schema/rolling_stock/light_rolling_stock.rs:61]: sorry, I didn't understand what needs updating?

File: editoast/src/views/infra/routes.rs (hunk 19 above)
Comment (hamz2a) [reply editoast/src/views/infra/routes.rs:39]: sorry, I didn't get what needs changing?

File: editoast/src/views/infra/errors.rs (hunk 18 above)
Comment (hamz2a) [reply editoast/src/views/infra/errors.rs]: could you please give a example?

File: editoast/src/models/rolling_stock/light_rolling_stock.rs (hunk 15 above)
Comment (leovalais) [reply editoast/src/models/rolling_stock/light_rolling_stock.rs]: Yeah but you made that field optional, so the `required:` object should be updated. This schema may be missing in the openapi.

File: editoast/src/schema/rolling_stock/light_rolling_stock.rs (hunk 17 above)
Comment (leovalais) [reply editoast/src/schema/rolling_stock/light_rolling_stock.rs:61]: You changed the type of the field so you should also update the type that diesel is expected to read.

File: editoast/src/views/infra/errors.rs (hunk 18 above)
Comment (leovalais) [reply editoast/src/views/infra/errors.rs]:
```suggestion
    #[diesel(sql_type = DieselJson<SchemaInfraError>)]
```

File: editoast/src/views/infra/objects.rs [hunk 21]
```diff
@@ -37,17 +38,16 @@ fn has_unique_ids(obj_ids: &Vec<String>) -> bool {
...
-    #[serde(skip_serializing)]
     obj_id: String,
     #[diesel(sql_type = Jsonb)]
     railjson: JsonValue,
     #[diesel(sql_type = Nullable<Jsonb>)]
-    geographic: Option<JsonValue>,
+    geographic: Option<DieselJson<GeoJson>>,
     #[diesel(sql_type = Nullable<Jsonb>)]
-    schematic: Option<JsonValue>,
+    schematic: Option<DieselJson<GeoJson>>,
```
Comment (leovalais) [inline editoast/src/views/infra/objects.rs:50]: We have a crate `postgis_diesel` to represent GeoJSON more precisely. `GeoJson` is currently only used for mvt tiles generation, but we've added an extern crate to represent GeoJSON values and we should remove this type. But that's for another PR

File: editoast/src/views/infra/routes.rs (hunk 19 above)
Comment (leovalais) [reply editoast/src/views/infra/routes.rs:39]: Nothing, I missclicked I guess 😅

File: editoast/src/views/layers/mod.rs [hunk 22]
```diff
@@ -131,7 +132,7 @@ async fn layer_view(
     Ok(Json(ViewMetadata {
         data_type: "vector".to_owned(),
         name: layer_slug.to_owned(),
-        promote_id: json!({layer_slug: layer.id_field}),
+        promote_id: HashMap::from([(layer_slug, layer.id_field.clone().unwrap_or_default())]),
```
Comment (leovalais) [inline editoast/src/views/layers/mod.rs:135]:
```suggestion
        promote_id: HashMap::from([(layer_slug, layer.id_field.unwrap_or_default())]),
```

If that conflicts with what's below you should unpack the struct beforehand.

**Review by leovalais — CHANGES_REQUESTED** [review]

File: editoast/src/models/rolling_stock/light_rolling_stock.rs (hunk 15 above)
Comment (hamz2a) [reply editoast/src/models/rolling_stock/light_rolling_stock.rs]: it was a mistake, I  will keep it as it was before, without `Option`.

File: editoast/src/schema/errors.rs (hunk 2 above)
Comment (leovalais) [reply editoast/src/schema/errors.rs]: It's not the tests' fault, it's serde's. flatten and deny_unknown_fields are incompatible (https://serde.rs/attr-flatten.html) so `schema::InfraError` cannot be deserialized properly, even if the serialization works. Changing how these errors are represented internally has to be discussed and is likely to impact things beyond the scope of this PR. Even more so since it overlaps with https://github.com/osrd-project/osrd/issues/4580#issuecomment-1673404336. So my suggestion would be to leave the JsonValue in the response of /infra/{id}/errors.

File: editoast/src/views/infra/errors.rs (hunk 18 above)
Comment (leovalais) [reply editoast/src/views/infra/errors.rs]: Cf https://github.com/osrd-project/osrd/pull/6200#discussion_r1434901650, this PR might not be the best place to address this change.

File: editoast/src/schema/rolling_stock/light_rolling_stock.rs (hunk 17 above)
Comment (leovalais) [reply editoast/src/schema/rolling_stock/light_rolling_stock.rs:61]: Nope my bad, this is fine, sorry 😖

File: editoast/src/views/infra/auto_fixes.rs [hunk 23]
```diff
@@ -284,6 +284,7 @@ mod test {
         // Check the only initial issues are "overlapping_speed_sections" warnings
         let infra_errors_before_all: PaginatedResponse<crate::views::infra::errors::InfraError> =
             read_body_json(call_service(&app, errors_request(small_infra_id)).await).await;
+
```
Comment (bougue-pe) [inline editoast/src/views/infra/auto_fixes.rs]: This file is going to move in other PRs, so to avoid useless conflicts, is it possible to avoid (fixup) that change?

File: editoast/src/views/infra/objects.rs [hunk 24]
```diff
@@ -155,6 +158,56 @@ mod tests {
...
+        let empty_infra_id = empty_infra.id();
+
+        let switch = Switch {
+            id: "switch_001".into(),
+            switch_type: "switch_type_001".into(),
+            ..Default::default()
+        }
+        .into();
+
+        let create_operation = Operation::Create(Box::new(switch));
```
Comment (bougue-pe) [inline editoast/src/views/infra/objects.rs]: minor: keeping a stricter type and convert only when needed
```suggestion
        let create_operation = Operation::Create(Box::new(switch.into()));
```

File: editoast/src/views/infra/objects.rs [hunk 25]
```diff
@@ -155,6 +158,56 @@ mod tests {
...
+                    "sncf": JsonValue::Null
+                },
+                "group_change_delay": 0.0,
+                "id": "switch_001",
+                "ports": {},
+                "switch_type": "switch_type_001"
+            }),
+            geographic: None,
+            schematic: None,
+        }];
```
Comment (bougue-pe) [inline editoast/src/views/infra/objects.rs]: minor:
this should be moved close to its usage IMO (in the `THEN` part).

also minor:
reusing defined strings for reuse/maintenance could be nicer (some other occurrences in tests may also be changed if agreed)
```suggestion
        let expected_switch_object = vec![ObjectQueryable {
            obj_id: switch.get_id().into(),
            railjson: json!({
                "extensions": {
                    "sncf": JsonValue::Null
                },
                "group_change_delay": 0.0,
                "id": switch.get_id(),
                "ports": {},
                "switch_type": switch.switch_type
            }),
            geographic: None,
            schematic: None,
        }];
```

**Review by bougue-pe — APPROVED** [review]

**Review by Tguisnet — APPROVED** [review]

**Review by flomonster — APPROVED** [review]
LGTM

**Review by leovalais — APPROVED** [review]
LGTM
